core/src/main/java/org/elasticsearch/index/mapper/Uid.java

```diff
@@ -123,4 +126,169 @@ public static String createUid(String type, String id) {
         return type + DELIMITER + id;
     }
 
+    private static final int UTF8 = 0xff;
+    private static final int NUMERIC = 0xfe;
+    private static final int BASE64_ESCAPE = 0xfd;
+
+    static boolean isURLBase64WithoutPadding(String id) {
+        // We are not lenient about padding chars ('=') otherwise
+        // 'xxx=' and 'xxx' could be considered the same id
+        final int length = id.length();
+        switch (length & 0x03) {
+        case 0:
+            break;
+        case 1:
+            return false;
+        case 2:
+            // the last 2 symbols (12 bits) are encoding 1 byte (8 bits)
+            // so the last symbol only actually uses 8-6=2 bits and can only take 4 values
+            char last = id.charAt(length - 1);
+            if (last != 'A' && last != 'Q' && last != 'g' && last != 'w') {
+                return false;
+            }
+            break;
+        case 3:
+            // The last 3 symbols (18 bits) are encoding 2 bytes (16 bits)
+            // so the last symbol only actually uses 16-12=4 bits and can only take 16 values
+            last = id.charAt(length - 1);
+            if (last != 'A' && last != 'E' && last != 'I' && last != 'M' && last != 'Q'&& last != 'U'&& last != 'Y'
```

Comment:
These conditionals (in case 2 above, and the allowed case below) could be optimized with boolean lookup tables.